src/dataflow-types/src/plan/mod.rs

```diff
@@ -397,6 +405,53 @@ impl Plan {
                     // Other plans are errors, and should be reported as such.
                     _ => return Err(()),
                 };
+                let mut required_arrangements = vec![HashSet::new(); inputs.len()];
+                // Delta joins should only be planned if a particular set of arrangements exists.
+                // Empirically, we've found that there are sometimes bugs causing them to be planned
+                // anyway. If this is the case, we need to create the arrangements, so we do so here,
+                // and complain with an error message.
+                if let JoinPlan::Delta(DeltaJoinPlan { path_plans }) = &plan {
+                    for DeltaPathPlan { stage_plans, .. } in path_plans {
+                        for DeltaStagePlan {
+                            lookup_relation,
+                            lookup_key,
+                            ..
+                        } in stage_plans
+                        {
+                            required_arrangements[*lookup_relation].insert(lookup_key.clone());
```

Comment:
Maybe performance golfing, but I bet we could skip the clone if we wanted (`difference` might need to be manually written, but .. also probably skips some more clones).

Reply:
I think that's right, but I don't think rewriting `difference` is worth the effort to save a few clones on a cold path.

Reply:
I can do it if you find it important, though.